### PR TITLE
Bug 535250 - Recursive annotations triggers a StackOverflowError

### PR DESCRIPTION
**Pull request by Tomas-Kraus**
Replaces PR 133.
Contains getAnnotation part of the fix and jUnit to verify self referencing meta-annotation and cycle of 3 annotations (A->B->C->A).

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=535250

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=535250

File: jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAnnotatedElement.java [hunk 1]
```diff
@@ -202,17 +204,36 @@ public MetadataAnnotation getAnnotation(Class annotation) {
...
+    protected final MetadataAnnotation getAnnotation(final String annotation, final Set<String> names) {
         if (m_annotations == null && m_metaAnnotations == null) {
             return null;
         }
-
         MetadataAnnotation metadataAnnotation = m_annotations.get(annotation);
         if (metadataAnnotation == null) {
             for (MetadataAnnotation a: m_metaAnnotations.values()) {
-                MetadataAnnotation ma = m_factory.getMetadataClass(a.getName()).getAnnotation(annotation);
+                if (names.contains(a.getName())) {
```
Comment (Tomas-Kraus) [inline jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAnnotatedElement.java]: @cristhiank I changed the code to always break the loop when there is a cycle. Unfortunately javax.persistence annotation won't work. Can you add some jUnit based on your own meta-annotations to verify this scenario?

Comment (cristhiank) [reply jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAnnotatedElement.java]: @Tomas-Kraus I can confirm that it is working now with non-existent annotations. I hope this fix gets merged soon. Thanks!

**Pull request merged**